### ISSUE TITLE
Unable to print transposed matrix when elements are not constant?

### ISSUE THREAD

**Issue by jyh1** [google-research/dex-lang#8]
The following example seems to work correctly:
```
mat1 = for i::4 j::4. 0
mat2.i.j = mat1.j.i
:p mat2
> [[0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0]]
```

However, weird error happened when the matrix contains different elements:
```
mat3 = for i::4 j::4. asint i
mat4.i.j = mat3.j.i
:p mat4
Compiler bug!Lookup failed:i
context:
[4=>4=>Int]
(tab::4=>Int = (for  j::4 . 
    mat3.j.i);
(tab_1::4=>4=>Int = (for  i::4 . 
    tab);
tab_1))
```

The type seems to be correctly inferred:
```
:t mat4
(4=>(4=>Int))
```

**Comment by dougalm** [comment]
Yep, that's a bug! Thanks for the report. These kinds of bugs can be a bit tricky to track down so it might have to wait till later in the week.

**Comment by dougalm** [comment]
Btw, to explain the error message a bit: that's a pretty-printing of the normalized IR which failed type checking because `i` in `mat3.j.i` is unbound. You can inspect the different IRs produced with `:typed`, `:normalized`, `:simp`, `:imp`, `:llvm`, and `:asm`.

**Comment by jyh1** [comment]
Not sure if this warrants a separate issue, another unbounded variable bug:

```
mat = for i::4 j::4 . asint i
:p for i. mat.i.i
[0, 1, 2, 3]

:p (tmp = for i. mat.i.i; tmp)
Compiler bug!Lookup failed:i
context:
[4=>Int]
mat.i
```
Feel like this is because the eta-reduction of AtomaticFor [here](https://github.com/google-research/dex-lang/blob/92a916859befc746fa050e55fb71b733d04d21ea/src/lib/Normalize.hs#L303) might be problematic... Should make sure `i` is not referenced in `e` before carrying out the reduction?

**Comment by dougalm** [comment]
Thanks for tracking that down! Even if it isn't the cause of this bug, you're right that the eta reduction is broken. Feel free to submit a PR. Otherwise I'll get to it soon.

**Comment by jyh1** [comment]
I would love to contribute! What are you thinking of the fix? Obviously we can simply turn off the eta reduction for now. To make it right requires information about free variables. I am still getting familiar with the codebase though I think currently there is no such information in the simplify pass?

**Comment by jyh1** [comment]
Fixed in #14.
